modules/ximgproc/src/quaternion.cpp

```diff
@@ -0,0 +1,206 @@
+﻿#include "precomp.hpp"
+using namespace std;
+using namespace cv;
+
+namespace cv
+{
+namespace ximgproc {
+
+void createQuaternionImage(InputArray _img, OutputArray _qimg)
+{
+    int type = _img.type(), depth = CV_MAT_DEPTH(type), cn = CV_MAT_CN(type);
+    CV_Assert((depth == CV_8U || depth == CV_32F || depth == CV_64F) && _img.dims() == 2 && cn == 3);
+    vector<Mat> qplane(4);
+    vector<Mat> plane;
+    split(_img, plane);
+    qplane[0] = Mat::zeros(_img.size(), CV_64FC1);
+    for (int i = 0; i < cn; i++)
+        plane[i].convertTo(qplane[3-i], CV_64F);
+    merge(qplane, _qimg);
+}
+
+void qconj(InputArray _img, OutputArray _qimg)
+{
+    int type = _img.type(), depth = CV_MAT_DEPTH(type), cn = CV_MAT_CN(type);
+    CV_Assert((depth == CV_32F || depth == CV_64F) && _img.dims() == 2 && cn == 4);
+    vector<Mat> qplane(4), plane;
+    split(_img, plane);
+    qplane[0] = plane[0].clone();
+    qplane[1] = -plane[1].clone();
+    qplane[2] = -plane[2].clone();
+    qplane[3] = -plane[3].clone();
+    merge(qplane, _qimg);
+}
+
+void qunitary(InputArray _img, OutputArray _qimg)
+{
+    int type = _img.type(), depth = CV_MAT_DEPTH(type), cn = CV_MAT_CN(type);
+    CV_Assert((depth == CV_64F) && _img.dims() == 2 && cn == 4);
+    vector<Mat> qplane(4), plane;
+    split(_img, plane);
+    qplane[0] = plane[0].clone();
+    qplane[1] = plane[1].clone();
+    qplane[2] = plane[2].clone();
+    qplane[3] = plane[3].clone();
+    double *ptr0 = qplane[0].ptr<double>(0, 0), *ptr1 = qplane[1].ptr<double>(0, 0);
+    double *ptr2 = qplane[2].ptr<double>(0, 0), *ptr3 = qplane[3].ptr<double>(0, 0);
+    int nb = plane[0].rows*plane[0].cols;
+    for (int i = 0; i < nb; i++, ptr0++, ptr1++, ptr2++, ptr3++)
+    {
+        double d = *ptr0 * *ptr0 + *ptr1 * *ptr1 + *ptr2 * *ptr2 + *ptr3 * *ptr3;
+        d = 1 / sqrt(d);
+        *ptr0 *= d;
+        *ptr1 *= d;
+        *ptr2 *= d;
+        *ptr3 *= d;
+    }
+    merge(qplane, _qimg);
+}
+
+void qdft(InputArray _img, OutputArray _qimg, int  	flags, bool sideLeft)
+{
+    //    CV_INSTRUMENT_REGION()
```

Comment:
It is OK to add this (into all exported "large enough" functions).
With ';' in the end.